File: framework/languages/java/e2e/SpotService/Shared/src/main/java/systems/zlink/e2e/spotservice/shared/EntryActorDestroyHandler.java
package systems.zlink.e2e.spotservice.shared;

import java.util.concurrent.CompletionStage;
import systems.zlink.framework.handlers.ZLinkSpotActorRequest;
import systems.zlink.framework.ZLinkMessageContext;

public final class EntryActorDestroyHandler {
    @ZLinkSpotActorRequest(packetName = "ActorDestroyReq")
    public CompletionStage<Contracts.ActorDestroyRes> handle(
        ScenarioEntrySpot spot,
        ScenarioActor actor,
        ZLinkMessageContext context,
        Contracts.ActorDestroyReq request) {
        if (!request.actorId().equals(actor.actorId())) {
            throw new IllegalStateException("destroy request actor does not match dispatched actor");
        }
        return spot.context().destroyActor(actor).thenApply(ignored -> {
            spot.record("ActorDestroyed", actor.actorId());
            return new Contracts.ActorDestroyRes(actor.actorId(), true);
        });
    }
}

File: framework/languages/java/e2e-kotlin/SpotService/Server/Session/src/main/kotlin/systems/zlink/e2e/kotlin/spotservice/session/handlers/EntryActorDestroyHandler.kt
package systems.zlink.e2e.kotlin.spotservice.session.handlers

import systems.zlink.e2e.kotlin.spotservice.Contracts
import systems.zlink.e2e.kotlin.spotservice.session.spots.ScenarioActor
import systems.zlink.e2e.kotlin.spotservice.session.spots.ScenarioEntrySpot
import kotlinx.coroutines.future.await
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.framework.handlers.ZLinkSpotActorRequest
import systems.zlink.framework.ZLinkMessageContext

class EntryActorDestroyHandler : ZLinkSuspendingEntrySpotActorRequestHandler<ScenarioEntrySpot, ScenarioActor, Contracts.DestroyActorReq, Contracts.DestroyActorRes> {
    @ZLinkSpotActorRequest(packetName = "DestroyActorReq")
    override suspend fun handle(
        spot: ScenarioEntrySpot,
        actor: ScenarioActor,
        context: ZLinkMessageContext,
        request: Contracts.DestroyActorReq,
    ): Contracts.DestroyActorRes {
        if (request.actorId != actor.actorId()) {
            throw IllegalStateException("destroy request actor does not match dispatched actor")
        }

        try {
            spot.context().runCpuWorker { true }.submit().await()
            spot.context().destroyActor(actor).await()
            spot.record("ActorDestroyed", actor.actorId())
        } catch (error: Throwable) {
            spot.record("ActorDestroyFailed", actor.actorId() + "/" + error.javaClass.simpleName)
        }

        return Contracts.DestroyActorRes(actor.actorId(), true)
    }
}
